lib/offline.js

```diff
@@ -63,52 +70,111 @@ module.exports = function(config) {
-        try {
-          stat = fs.statSync(path.join(rootDir, script));
-        } catch (ex) {
-          console.log(chalk.red.bold(script + ' doesn\'t exist.'));
-          throw ex;
-        }
-
-        if (!stat.isFile()) {
-          console.log(chalk.red.bold(script + ' is not a file.'));
-          throw new Error(script + ' is not a file.');
-        }
+    function getFilesAndHashes(files, rootDir, sizeWhiteList) {
+      var totalSize = 0;
+      var filesAndHashes = files.map(function (filepath) {
+        totalSize += countFileSize(filepath, sizeWhiteList);
+        var data = fs.readFileSync(filepath);
+        var hash = getHash(data);
+        logOk(filepath);
+        return {
+          path: filepath.replace(rootDir, function (match, offset) {
+            //XXX: The root must be the worker's directory
+            return offset === 0 ? './' : match;
+          }),
+          hash: hash,
+        };
       });
 
-      return swPrecache.write(path.join(rootDir, 'offline-worker.js'), {
-        staticFileGlobs: staticFileGlobs,
-        stripPrefix: rootDir,
-        verbose: true,
-        logger: function(message) {
-          if (message.indexOf('Caching') === 0) {
-            message = chalk.bold.green('✓ ') + message.replace('static resource ', '');
-          }
-
-          console.log(message);
-        },
-        importScripts: importScripts,
-        cacheId: cacheId,
-        ignoreUrlParametersMatching: config.ignoreUrlParametersMatching || [/./],
-        maximumFileSizeToCacheInBytes: Infinity,
+      console.log('Total precache size is about ' + prettyBytes(totalSize) +
+                  ' for ' + files.length + ' resources.');
+
+      return filesAndHashes;
+    }
+
+    function checkImports(imports, rootDir) {
+      imports.forEach(function (filepath) {
+        //TODO: I think we should warn, not fail
+        assertExists(path.join(rootDir, filepath));
       });
-    }));
+    }
+
+    function pluckHashes(entries) {
+      return entries.map(function (entry) { return entry.hash; });
+    }
+
+    function assertExists(filepath) {
+      var stat;
+      try {
+        stat = fs.statSync(filepath);
+      } catch (ex) {
+        console.log(chalk.red.bold(filepath + ' doesn\'t exist.'));
+        throw ex;
+      }
+
+      if (!stat.isFile()) {
+        console.log(chalk.red.bold(filepath + ' is not a file.'));
+        throw new Error(filepath + ' is not a file.');
+      }
+    }
+
+    function countFileSize(filepath, whiteList) {
+      var stat = fs.statSync(filepath);
+      if (stat.isFile() && stat.size > 2 * 1024 * 1024 && whiteList.indexOf(filepath) === -1) {
+        console.log(chalk.yellow.bold(filepath + ' is bigger than 2 MiB. Are you sure you want to cache it? To suppress this warning, explicitly include the file in the fileGlobs list.'));
+      }
+      return stat.size;
+    }
+
+    function logOk(filepath) {
+      console.log(chalk.bold.green('✓ ') + 'Caching ' + filepath);
```

Comment:
This doesn't put quotes around the file path, as the current implementation does, but I like this version better, since the quotes don't add much, as it's already easy to distinguish the path from the word "Caching" that appears before it.

Also missing here is the size of the file in parentheses after the path. I have mixed feelings about that change, but I'm ok with it, as it isn't clear how useful that information is.